### ISSUE TITLE
Implement Number

### ISSUE THREAD

**Issue by tadfisher** [ionspin/kotlin-multiplatform-bignum#170]
**Is your feature request related to a problem? Please describe.**

I'm trying to serialize a `BigDecimal` using `kotlinx.serialization` into a JSON-formatted string,** but the library gives me two choices:

1. Convert to a Double before serializing, resulting in IEEE-754-induced precision loss
2. Serialize as a quoted string, breaking compatibility with the receiving deserializer

**Describe the solution you'd like**

`kotlinx.serialization` does support the `kotlin.Number` type, which does exactly what I want: call `toString()` on the value and encode the resulting string as an unquoted JSON number.

**Describe alternatives you've considered**

The other option for me is to add `expect class BigDecimal : Number` and essentially reimplement this library.

**Additional context**

Bug report about serializing bignums: https://github.com/Kotlin/kotlinx.serialization/issues/1051

**Comment by tadfisher** [comment]
Here's a hack I found to make this work:

```kotlin
// HACK: Wrap BigDecimal to extend Number, allowing us to serialize as an unquoted string.
private class BigDecimalNumber(val value: BigDecimal) : Number() {
    override fun toByte(): Byte = throw UnsupportedOperationException()
    override fun toChar(): Char = throw UnsupportedOperationException()
    override fun toDouble(): Double = throw UnsupportedOperationException()
    override fun toFloat(): Float = throw UnsupportedOperationException()
    override fun toInt(): Int = throw UnsupportedOperationException()
    override fun toLong(): Long = throw UnsupportedOperationException()
    override fun toShort(): Short = throw UnsupportedOperationException()
    override fun toString(): String = value.toStringExpanded()
}
```

**Comment by ionspin** [comment]
Hi @tadfisher ,

I actually wanted to implement serialization long time ago, it just slipped my mind. Thanks for the feature request and the workaround, I'll see if I can write a proper solution over the weekend!

**Comment by volkert-fastned** [comment]
@tadfisher Thanks for the workaround! I believe the precision loss on serialization can be avoided when we make use of the (currently still experimental) `JsonUnquotedLiteral` instead of `JsonPrimitive`, which would also eliminate the need for the `BigDecimalNumber`. This is done by rewriting the `override fun serialize` part as follows:

```kotlin
    override fun serialize(encoder: Encoder, value: BigDecimal) {
        val element = JsonUnquotedLiteral(value.toStringExpanded())
        (encoder as JsonEncoder).encodeJsonElement(element)
    }
```

@ionspin I'm not sure what "additional dependencies" you are referring to. With the above modification, @tadfisher's workaround only needs the following mult-platform imports:

```kotlin
import com.ionspin.kotlin.bignum.decimal.BigDecimal
import com.ionspin.kotlin.bignum.decimal.toBigDecimal
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializer
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonUnquotedLiteral
```

**Comment by ionspin** [comment]
I am referring to kotlinx.serialization library.

**Comment by volkert-fastned** [comment]
@ionspin OK, but then it woul be useful to include this serializer in the `bignum-serialization-kotlinx` library, right? that library/dependency is already dependent on kotlinx.serialization.

**Comment by ionspin** [comment]
Yes, that would be great, pull request would be highly appreciated!

**Comment by volkert-fastned** [comment]
Would be happy to. But `bignum-serialization-kotlinx` doesn't appear to have its own repository. Is that correct?

**Comment by ionspin** [comment]
Yeah, that is correct.
